### PR TITLE
feat: new version of DID resolver + updated DID method specification

### PR DESCRIPTION
**Pull request by Diiaablo95**
# New DIF Universal Resolver KILT plugin

This PR fixes KILTProtocol/ticket#1296 and KILTProtocol/ticket#1297.

The updated spec follows the guidelines in the [DID Core Spec](https://www.w3.org/TR/did-core/#methods) for new method specifications.

For reference, please check the [DIF GitHub repo](https://github.com/decentralized-identity/universal-resolver), and especially the [driver development guide](https://github.com/decentralized-identity/universal-resolver/blob/main/docs/driver-development.md).

Once this PR is merged, I will make a new release, which will also push a new image to DockerHub, and open a PR in the DIF repo to update the driver.

### CODE REVIEW

File: docs/did-spec/spec.md [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+</table>
+
+If the first case where `deactivated` is true, the resulting `didDocument` will be a minimal document containing only the `id` and `@context` properties as specified in the DID Core specification.
+
+On the KILT blockchain, information related to a full DID is stored under multiple storage maps within the `did` pallet, specifically:
+
+- `did -> did(AccountId32): Option<DidDetails>`: maps from a DID identifier to its details, if present.
+An instance of `DidDetails` contains all the keys under the control of the DID subject.
+- `did -> didBlacklist(AccountId32): Option<()>`: maps from a DID identifier to an optional empty tuple indicating whether a given DID has been deleted (optional tuple is not null) or not (optional tuple is null).
+- `did -> serviceEndpoints(AccountId32, Vec<u8>): Option<DidEndpoint>`: maps from the concatenation of hashed DID identifier and hashed service ID to the service endpoints details, if present.
```
Comment (csmarc) [inline docs/did-spec/spec.md:214]:
```suggestion
- `did -> serviceEndpoints(AccountId32, Vec<u8>): Option<DidEndpoint>`: maps from the concatenation of hashed DID identifier and hashed service ID to the details of a service endpoint, if present.
```

File: docs/did-spec/spec.md [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+Upon lookups, by only providing the first key, it is possible to retrieve all the service IDs under a given DID identifier.
+An instance of `DidEndpoint` contains the service details, i.e., service ID, a set of service types, and a set of service URLs.
+
+> The validity and authenticity of the information stored on the blockchain are guaranteed by the same blockchain, which verifies that all DID management operations other than creation are signed by the DID authentication key, as described in the section relative to each operation.
```
Comment (csmarc) [inline docs/did-spec/spec.md:218]: > as described in the section relative to each operation.

What do you mean here? Respective sections in the docs?

File: docs/did-spec/spec.md [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+To incentivise users to only keep the relevant data on the KILT blockchain, writing a DID requires locking up some KILTs of deposit which is then returned when the DID is deleted.
+Since DIDs are decoupled from KILT accounts, a DID needs a KILT account with enough funds to pay for the creation deposit.
+In order not to make the deposit payer dependent on the DID subject to delete the DID and obtain the deposit back, the KILT DID method also allows the deposit payer to delete the DID information from the blockchain and get the deposit refunded.
+
+Hence, a DID creation operation requires the DID subject to explicitly specify the authorised submitter of the operation as part of the signed data.
+This ensures that nobody can submit the DID creation operation on behalf of the designated user and, consequently, that nobody can delete a DID without the initial authorisation of the DID subject expressed as part of the creation operation signature.
+
+## Denial-of-Service
+
+The almost totality of extrinsics are subject to a fee, comparable to the concept of gas in the Ethereum ecosystem, which has real-world monetary value.
```
Comment (csmarc) [inline docs/did-spec/spec.md:309]:
```suggestion
The blockchain transactions (extrinsics) are subject to a fee depending on their raw length, on-chain computational intensity and storage requirements, similarly to the concept of gas in the Ethereum ecosystem, which represents real-world monetary value.
```

File: docs/did-spec/spec.md [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+Hence, attempts to spam the blockchain with bogus transactions will either be blocked by a node before being included in a block, thus reducing the effect of such action, or they will result in an error but in the payment of a dispatch fee which, eventually, will lead to resource exhaustion for the attacker's account.
```
Comment (csmarc) [inline docs/did-spec/spec.md:310]:
```suggestion
Hence, attempts to spam the blockchain with bogus transactions will either be blocked by a node before being included in a block, thus reducing the effect of such an attack, or they will result in an error while the attacker still needs to pay of a transaction dispatch fee which, eventually, would exhaust the attacker resources.
```

**Review by csmarc — COMMENTED** [review]
DID Spec looks really good 👌  Some rephrasing suggestions provided and typo fixes.

File: docs/did-spec/spec.md [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+A KILT full DID has the following structure:
+
+```
+kilt-did            = "did:kilt:"<did-identifier>
+did-identifier      = <base-58-encoded-kilt-address>
+```
+
+where `<base-58-encoded-kilt-address>` is a KILT address.
+
+A full DID relies on the [KILT blockchain][spiritnet-polkadot] for management and resolution, it allows the storage of more than one authentication and one key agreement key, and allows the DID subject to perform other types of operations, e.g., issuing credentials, on the KILT blockchain.
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply docs/did-spec/spec.md:86]: Nope. Creation of a credential happens on chain, and the attestation key must be present on chain, aka, the DID must be a full DID.

[the remark this answers is not in the document]
Comment (Diiaablo95) [reply docs/did-spec/spec.md:86]: I am not 100% the comma should be removed here. Issuing credential is one example of operation that can be performed on the KILT blockchain.

File: docs/did-spec/spec.md (hunk 2 above)
Comment (Diiaablo95) [reply docs/did-spec/spec.md:218]: Yeah I was referring to the sections in the docs that describe those other operations, for instance DID update or DID deletion.

File: docs/did-spec/spec.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (csmarc) [reply docs/did-spec/spec.md:86]: ahh, yes, I see

File: docs/did-spec/spec.md (hunk 2 above)
Comment (Dudleyneedham) [reply docs/did-spec/spec.md:218]: Can we say the KILT blockchain?

**Review by Dudleyneedham — COMMENTED** [review]
One minor comment. Overall, no nitpicks from me.

**Review by tjwelde — APPROVED** [review]
LGTM

**Pull request merged**